### ISSUE TITLE
Update Gradle configuration in a flutter module project

### ISSUE THREAD

**Issue by joshuadeguzman** [flutter/flutter#36185]
## Overview
Given that the `.android` (and `.ios`) are autogenerated folders, how are we supposed to update the Gradle configuration (eg. `minSdkVersion` outputted in the `./android/build.gradle` file)?

I encountered the problem when using `flutter_secured_storage`, however, I believe this is not the problem of the plugin maintainer, but rather flutter's constraints from letting us to configure it ourselves (updating files inside `.android/build.gradle`). 

> NOTE: It's `.android` (autogerated dir), not the defined `android` package from the normal flutter templates.

In a regular flutter project template, I will just update the `minSdkVersion` inside `android/build.gradle` configuration. 

However, how are we supposed to do it in a flutter module?

> NOTE: I included flutter as a submodule in my native android app repo. Running `flutter packages get` will re-create `.android` dir. That's why applied changes in it will be disregarded.

## Logs

`flutter run -v`
```
 uses-sdk:minSdkVersion 16 cannot be smaller than version 18 declared in library [:flutter_secure_storage] /Users/joshuadeguzman/Documents/Tools/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_secure_storage-3.2.1+1/android/build/intermediates/merged_manifests/debug/processDebugManifest/merged/AndroidManifest.xml as the library might be
using APIs not available in 16
[        ]      Suggestion: use a compatible library with a minSdk of at most 16,
[        ]              or increase this project's minSdk version to at least 18,
[        ]              or use tools:overrideLibrary="com.it_nomads.fluttersecurestorage" to force usage (may lead to runtime failures)
[        ] FAILURE: Build failed with an exception.
```

`flutter analyze`
```
Running "flutter pub get" in flutter_code...                        1.7s
Analyzing flutter_code...
No issues found! (ran in 6.7s)
```

`flutter doctor -v`
```
[✓] Flutter (Channel master, v1.8.2-pre.33, on Mac OS X 10.14.4 18E224, locale en-PH)
    • Flutter version 1.8.2-pre.33 at /Users/joshuadeguzman/Documents/Tools/flutter
    • Framework revision e3ee5c6bbb (35 hours ago), 2019-07-13 16:02:09 -0700
    • Engine revision 919e353236
    • Dart version 2.5.0 (build 2.5.0-dev.1.0 24725a8559)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
    • Android SDK at /Users/joshuadeguzman/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
    ✗ CocoaPods installed but not initialized.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To initialize CocoaPods, run:
        pod setup
      once to finalize CocoaPods' installation.

[✗] iOS tools - develop for iOS devices
    ✗ libimobiledevice and ideviceinstaller are not installed. To install with Brew, run:
        brew update
        brew install --HEAD usbmuxd
        brew link usbmuxd
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    ✗ ios-deploy not installed. To install:
        brew install ios-deploy

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 36.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.36.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.2.0

[✓] Connected device (3 available)
    • Pixel XL • HT6C90202587 • android-arm64  • Android 8.0.0 (API 26)
    • macOS    • macOS        • darwin-x64     • Mac OS X 10.14.4 18E224
    • Chrome   • chrome       • web-javascript • Google Chrome 75.0.3770.100


! Doctor found issues in 2 categories.
```

NOTE: Disregard iOS issues, I'm currently targeting Android builds.

Thanks for the help!

**Comment by eseidelGoogle** [comment]
@blasten might know.

**Comment by blasten** [comment]
There isn't a way currently.  Tentatively, we could make this configurable in the module's `pubspec.yaml` in the same way `androidX` and `androidPackage` are.  We talked about generating the `.android` and `.ios` at flutter create time, just like an app project, but we haven't fully evaluated the implications. cc @xster

**Comment by xster** [comment]
Added a note in https://github.com/flutter/flutter/issues/32862 to document it. 

@blasten I assume you meant there's no way in gradle to specify a midSdkVersion as an actual minimum for the present AAR but let downstream dependencies specify a higher minSdkVersion?

Otherwise, it's a scenario we should consider when deciding on https://github.com/flutter/flutter/issues/32989. For now, @joshuadeguzman you can run `flutter make-host-app-editable` and edit the gradle file. But you'd be on the hook for 'updating' that file in the future (which could just be `rm -rf android/`, `flutter packages get`).

**Comment by joshuadeguzman** [comment]
Thanks for looking into this.

As per suggestion, yes I did use `flutter make-host-app-editable`, and it works. On a note, the project `build.gradle` of the generated `android` folder is still not editable, given that not much configuration is needed for that (except when updating the android gradle plugin to `3.3.x`, etc.), this works fine for now.

**Comment by xster** [comment]
> On a note, the project build.gradle of the generated android folder is still not editable 

What did you mean by it? Can you point to the exact file path?

**Comment by joshuadeguzman** [comment]
Hi @xster, it's on `android/build.gradle`. 

Has the file header `// Generated file. Do not edit.`

**Comment by rvail2** [comment]
I'm having a similar issue. If I change dependencies in pubspec.yaml and run 'flutter pub get', I lose a couple changes.  My alteration to use 'com.android.tools.build:gradle:3.4.2' is reverted back to 3.2.1, along with the distribution url in my gradle-wrapper. (from 5.1.1 to 4.6).  If I don't remember to manually update these, the module breaks our native app.

**Comment by joshuadeguzman** [comment]
Hi @rvail2, have you tried using `flutter make-host-app-editable`?

**Comment by xster** [comment]
Ha, I guess the problem is that our newly generated non-ephemeral templates didn't remove that line of comment on top of the file. It shouldn't affect anything (i.e. that comment isn't right). 

It's a small clean up we'd likely want to do after #32989

**Comment by rvail2** [comment]
@joshuadeguzman sorry for the late reply.  I have not.  I will try as soon as I get the chance.

**Comment by blasten** [comment]
This issue is a duplicate of https://github.com/flutter/flutter/issues/32989

**Comment by svenjacobs** [comment]
I'm having the same issue trying to add the `camera` plugin to a Flutter module in an "add 2 app" situation. `camera` requires minSDK 21 however the generated project in `.android` declares minSDK 16.

In [this article](https://flutter.dev/docs/development/add-to-app/android/plugin-setup) Google clearly states that `.android` should not be modified or added to VCS. They even mention the `camera` module as an example. However this does not works and can be reproduced with a few simple steps:

 1. Create a new module: `flutter create --template=module --project-name example`
 2. Add `camera` plugin to `pubspec.yaml`
 3. `flutter pub get`
 4. Now running `flutter build apk` results in the following error:

```
/Users/sven/Development/example/.android/app/src/main/AndroidManifest.xml Error:
        uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [:camera] /Users/sven/.pub-cache/hosted/pub.dartlang.org/camera-0.5.8+11/android/build/intermediates/library_manifest/release/AndroidManifest.xml as the library might be using APIs not available in 16
        Suggestion: use a compatible library with a minSdk of at most 16,
                or increase this project's minSdk version to at least 21,
                or use tools:overrideLibrary="io.flutter.plugins.camera" to force usage (may lead to runtime failures)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processReleaseManifest'.
> Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [:camera] /Users/sven/.pub-cache/hosted/pub.dartlang.org/camera-0.5.8+11/android/build/intermediates/library_manifest/release/AndroidManifest.xml as the library might be using APIs not available in 16
        Suggestion: use a compatible library with a minSdk of at most 16,
                or increase this project's minSdk version to at least 21,
                or use tools:overrideLibrary="io.flutter.plugins.camera" to force usage (may lead to runtime failures)
```

Flutter version **1.22.4**.  

Running `flutter make-host-app-editable` results in the following output:

```
[!] The "make-host-app-editable" command is deprecated and will be removed in a future version of Flutter. See https://flutter.dev/docs/development/tools/sdk/releases for previous
releases of Flutter.
```

How are we supposed to add plugins to a Flutter module that require adjustments to the generated `.android` project? It's not just the minSDK. `firebase_ml_vision` for instance requires further changes like initializing a Firebase project...

**Comment by xster** [comment]
(generally, we're not monitoring closed issues. Please open new issues to get responses)

@svenjacobs, in the above example, the failure is from the attached sample wrapper project included with the module but it's just for illustration purposes. The sample wrapper project shouldn't be "used" in a production setting. To "use" the module, include it in your own wrapper project where you can set the min sdk version to any as needed.

**Comment by svenjacobs** [comment]
> (generally, we're not monitoring closed issues. Please open new issues to get responses)
> 
> @svenjacobs, in the above example, the failure is from the attached sample wrapper project included with the module but it's just for illustration purposes. The sample wrapper project shouldn't be "used" in a production setting. To "use" the module, include it in your own wrapper project where you can set the min sdk version to any as needed.

@xster Thanks for the reply and the suggested solution. I would appreciate if the documentation is updated because currently it sounds like modules and plugins like `camera` work out-of-the-box, which is not the case. Thank you! 😄
